internal/server/boltdbstate/job.go

```diff
@@ -1709,6 +1724,147 @@ func (s *State) taskComplete(jobId string) error {
+// and if so, progress the PipelineRun state machine.
+func (s *State) pipelineAck(jobId string) error {
+	job, err := s.JobById(jobId, nil)
+	if err != nil {
+		s.log.Error("error getting job by id", "job", jobId, "err", err)
+		return err
+	} else if job.Pipeline == nil {
+		s.log.Trace("job is not part of a pipeline", "job", jobId)
+		return nil
+	}
+
+	// grab pipeline run that triggered the job based on the PipelineTask Ref
+	run, err := s.PipelineRunGet(&pb.Ref_Pipeline{
+		Ref: &pb.Ref_Pipeline_Id{
+			Id: &pb.Ref_PipelineId{
+				Id: job.Pipeline.Pipeline,
+			},
+		},
+	}, job.Pipeline.RunSequence)
+	if err != nil || len(run.Jobs) < 1 {
+		s.log.Error("failed to retrieve pipeline to ack", "job", job.Id, "pipeline id", job.Pipeline.Pipeline, "run", job.Pipeline.RunSequence)
+		return err
+	}
+
+	for i, j := range run.Jobs {
+		// If job ID matches the job queued by the pipeline run
+		if j.Id == job.Id {
+			// first job is the start job
+			if i == 0 {
+				run.Status = pb.PipelineRun_STARTING
+				s.log.Trace("pipeline run is starting", "job", job.Id, "pipeline", job.Pipeline.Pipeline, "run", run.Sequence)
+			} else {
+				// any other queued job indicates the pipeline is running
+				run.Status = pb.PipelineRun_RUNNING
+				s.log.Trace("pipeline is running", "job", job.Id, "pipeline", job.Pipeline.Pipeline, "run", run.Sequence)
+			}
+		} else {
+			return status.Errorf(codes.Internal, "no job queued by pipeline %q run %q matches the requested job id %q", job.Pipeline.Pipeline, job.Pipeline.RunSequence, job.Id)
+		}
+	}
+
+	// PipelineRunPut the new state
+	if err := s.PipelineRunPut(run); err != nil {
+		s.log.Error("failed to ack pipeline run state", "job", job.Id, "pipeline", job.Pipeline.Pipeline, "run", job.Pipeline.RunSequence)
+		return err
+	}
+
+	return nil
+}
+
+// pipelineCancel will look up the referenced job to see if it has a PipelineStep ref
+// associated with it, and if so, progress the pipelineRun state machine.
+func (s *State) pipelineCancel(jobId string) error {
+	job, err := s.JobById(jobId, nil)
+	if err != nil {
+		s.log.Error("error getting job by id", "job", jobId, "err", err)
+		return err
+	} else if job.Pipeline == nil {
+		s.log.Trace("job is not part of a pipeline", "job", jobId)
+		return nil
+	}
+
+	// grab the pipeline run
+	run, err := s.PipelineRunGet(&pb.Ref_Pipeline{
+		Ref: &pb.Ref_Pipeline_Id{
+			Id: &pb.Ref_PipelineId{
+				Id: job.Pipeline.Pipeline,
+			},
+		},
+	}, job.Pipeline.RunSequence)
+	if err != nil || len(run.Jobs) < 1 {
+		s.log.Error("failed to retrieve pipeline to complete", "job", job.Id, "pipeline", job.Pipeline.Pipeline, "run", job.Pipeline.RunSequence)
+		return err
+	}
+
+	// mark pipeline run state as cancelled
+	for _, j := range run.Jobs {
+		if j.Id == job.Id {
+			if job.State == pb.Job_ERROR || job.State == pb.Job_SUCCESS {
+				return nil
+			} else {
+				run.Status = pb.PipelineRun_CANCELLED
```

Comment:
We shouldn't need to loop over any of these jobs, I'd mark the whole run as cancelled if any job in a pipeline run gets cancelled.

Reply:
Ahh, the job system will cancel any dependent jobs when a job is cancelled I believe. I don't think we have a "continue on failure" in the job system that I know of!

Reply:
https://github.com/hashicorp/waypoint/blob/main/internal/server/boltdbstate/job.go#L927

Reply:
& the test suggests the behavior is if the job is already complete, cancel does nothing and doesn't affect dependents
https://github.com/hashicorp/waypoint/blob/main/pkg/serverstate/statetest/test_job.go#L2113

Reply:
This is resolved per in-person discussion :)
But brings up another problem... in our current job system, we change job status to `ERROR` whether it errored or was manually cancelled. The timeout cancellation (error) calls the same 'JobCancel' method as the manual cancel flow, which then calls `pipelineCancel` after `jobCancel` is successful. So there's no easy way to set pipeline state = error based on job state or how the cancellation flow is called. We could put `pipelineCancel` before `jobCancel` and filter for error state jobs? But I'm leaning toward just having a "cancelled" state rather than separate "error" and "cancel".